File: objective-c/test/Ice/interceptor/InterceptorI.m
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#import <interceptor/InterceptorI.h>
#import <InterceptorTest.h>
#import <TestCommon.h>

@implementation InterceptorI

-(id) init:(ICEObject*) servant_
{
    self = [super init];
    if(!self)
    {
        return nil;
    }
    servant = ICE_RETAIN(servant_);
    return self;
}

#if defined(__clang__) && !__has_feature(objc_arc)
-(void) dealloc
{
    [lastOperation release];
    [servant release];
    [super dealloc];
}
#endif

-(void) dispatch:(id<ICERequest>) request
{
    ICECurrent* current = [request getCurrent];

    ICE_RELEASE(lastOperation);
    lastOperation = ICE_RETAIN(current.operation);
    lastStatus = NO;

    if([lastOperation isEqualToString:@"addWithRetry"])
    {
        int i = 0;
        for(i = 0; i < 10; ++i)
        {
            @try
            {
                [servant ice_dispatch:request];
                test(NO);
            }
            @catch(TestInterceptorRetryException*)
            {
                //
                // Expected, retry
                //
            }
        }

        [(NSMutableDictionary*)current.ctx setObject:@"no" forKey:@"retry"];

        //
        // A successful dispatch that writes a result we discard
        //
        [servant ice_dispatch:request];
    }

    [servant ice_dispatch:request];
    lastStatus = YES;
}

-(BOOL) getLastStatus
{
    return lastStatus;
}

-(NSString*) getLastOperation
{
    return lastOperation;
}

-(void) clear
{
    lastStatus = NO;
    ICE_RELEASE(lastOperation);
    lastOperation = nil;
}

@end

File: java-compat/test/src/main/java/test/Ice/interceptor/InterceptorI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package test.Ice.interceptor;

import test.Ice.interceptor.Test.RetryException;
import test.Ice.interceptor.Test.InvalidInputException;

class InterceptorI extends Ice.DispatchInterceptor
{
    InterceptorI(Ice.Object servant)
    {
        _servant = servant;
    }

    protected static void
    test(boolean b)
    {
        if(!b)
        {
            throw new RuntimeException();
        }
    }

    @Override
    public boolean
    dispatch(Ice.Request request)
        throws Ice.UserException
    {
        Ice.Current current = request.getCurrent();

        String context = current.ctx.get("raiseBeforeDispatch");
        if(context != null)
        {
            if(context.equals("user"))
            {
                throw new InvalidInputException();
            }
            else if(context.equals("notExist"))
            {
                throw new Ice.ObjectNotExistException();
            }
            else if(context.equals("system"))
            {
                throw new MySystemException();
            }
        }

        _lastOperation = current.operation;

        if(_lastOperation.equals("addWithRetry"))
        {
            for(int i = 0; i < 10; ++i)
            {
                try
                {
                    _servant.ice_dispatch(request);
                    test(false);
                }
                catch(RetryException re)
                {
                    //
                    // Expected, retry
                    //
                }
            }

            current.ctx.put("retry", "no");
        }

        _lastStatus = _servant.ice_dispatch(request);

        context = current.ctx.get("raiseAfterDispatch");
        if(context != null)
        {
            if(context.equals("user"))
            {
                throw new InvalidInputException();
            }
            else if(context.equals("notExist"))
            {
                throw new Ice.ObjectNotExistException();
            }
            else if(context.equals("system"))
            {
                throw new MySystemException();
            }
        }

        return _lastStatus;
    }

    boolean
    getLastStatus()
    {
        return _lastStatus;
    }

    String
    getLastOperation()
    {
        return _lastOperation;
    }

    void
    clear()
    {
        _lastOperation = null;
        _lastStatus = false;
    }

    protected final Ice.Object _servant;
    protected String _lastOperation;
    protected boolean _lastStatus;
}
